glean-core/ios/Glean/Dispatchers.swift

```diff
@@ -153,15 +157,27 @@ class Dispatchers {
 
     /// Stop queuing tasks and process any tasks in the queue.
     func flushQueuedInitialTasks() {
-        // Add all of the queued operations to the `operationQueue` which will cause them to be
-        // executed serially in the order they were collected.  We are passing `testingMode` to the
-        // `waitUntilFinished` parameter since this is a serial queue and any newly queued tasks
-        // should execute after the `preInitOperations` that are being added here. For tests, we
-        // need to await all of the tasks to finish execution, so we set this to true.
-
-        // TODO(bug 1591094) Android has a timeout of 5 seconds for running preinit tasks.
+        // Timeouts are easily accomplished in Swift using DispatchSemaphores and launching
+        // a block of code asynchronously.  This creates a semaphore that we can await with a
+        // timeout to prevent the queued initial tasks from hanging up or taking a long time.
+        let timeout = DispatchSemaphore(value: 0)
+        concurrentOperationsQueue.addOperation {
+            // Add all of the queued operations to the `operationQueue` which will cause them to be
+            // executed serially in the order they were collected.
+            self.serialOperationQueue.addOperations(self.preInitOperations, waitUntilFinished: true)
+            timeout.signal()
+        }
 
-        self.serialOperationQueue.addOperations(preInitOperations, waitUntilFinished: testingMode)
+        // Await the async task to complete up to the allowed time and log an error if it
+        // times out.
+        if timeout.wait(timeout: DispatchTime.now() + Constants.queueProcessingTimeout) == .timedOut {
+            logger.error("Timeout processing initial tasks queue")
+            // Turn off queuing to allow for normal background execution mode
+            queueInitialTasks.value = false
```

Comment:
I had to duplicate this line here, otherwise the recording of the error on the next line would get lost when the queue was dumped.